### DISCUSSION TITLE
My ComfyUI beginners questions

### DISCUSSION THREAD

**Discussion by MoonMoon82** [comfyanonymous/ComfyUI#74] [Q&A]
I'm a automatic1111-webui and Blender user. So it was quite easy for me to get into ComfyUI <3

But I still have some questions I couldn't/barely figure out it this repository and through web searches:

1. Face restoration? Is it already in there? Any examples? CodeFormer-Support?
2. Is it possible to create and save own node groups (like in Blender) ?
3. Is there any chance to see the intermediate images during the calculation of a sampler node (like in 1111 WebUI settings "Show new live preview image every N sampling steps") ?
4. Any way to integrate external text files as input for CLIPTextEncoder nodes?
5. How do I change conditions depending on the sampling cycle? 
6. ControlNet: There's an "Guess Mode" in 1111 WebUI which I could figure out in ComfyUI - any chances?
7. ControlNet: In 1111 WebUI ControlNet has "Guidance Start/End (T)" sliders. I guess it refers to my 5th question. Is there any equivalent in ComfyUI ?
8. ControlNet: Where are the preprocessors which are used to feed controlnet models?

So far, great work, awesome project!

**Comment by comfyanonymous** [comment]
>Face restoration? Is it already in there? Any examples? CodeFormer-Support?

Not yet. There's a chance the model might work with the regular upscale with model node but it probably needs a proper node added.

>Is it possible to create and save own node groups (like in Blender) ?

Not yet.

>Is there any chance to see the intermediate images during the calculation of a sampler node (like in 1111 WebUI settings "Show new live preview image every N sampling steps") ?

The KSamplerAdvanced node can be used to sample on an image for a certain number of steps but if you want live previews that's "Not yet."

>How do I change conditions depending on the sampling cycle?

By chaining KSamplerAdvanced nodes a bit like these slightly complex examples: https://comfyanonymous.github.io/ComfyUI_examples/noisy_latent_composition/

By using KSamplerAdvanced you can change models/samplers/cfg/controlnets/whatever settings mid way through the generation.


>ControlNet: Where are the preprocessors which are used to feed controlnet models?

They are implemented as custom nodes here: https://github.com/Fannovel16/comfy_controlnet_preprocessors

**Reply by MoonMoon82** [reply]
@comfyanonymous Thank you very much for your feedback to my questions and sorry for my impatience!
Is there an overview/index of popular custom nodes? I've already seen some custom nodes mentioned in issues/feature requests. But it's quite difficult to find out if a custom node already exists.

**Comment by PladsElsker** [comment]
> ControlNet: In 1111 WebUI ControlNet has "Guidance Start/End (T)" sliders. I guess it refers to my 5th question. Is there any equivalent in ComfyUI ?  

It's possible to use chained KSamplerAdvanced nodes to offset the start and end of controlNets

**Reply by MoonMoon82** [reply]
I already tried it with chained sampler nodes. Quite horrifying complexity the nodesetup becomes if you try to reconstruct some basic Guidance Start-End points and/or simple [start:end:steps] prompt expressions. I hope there will be an alternative in the future :/

**Reply by PladsElsker** [reply]
I agree, this can become quite the setup...

**Reply by DingSiuyo** [reply]
Hi, could you please give an example? Cheers

**Reply by PladsElsker** [reply]
sure! Here's an example that starts with no controlnet from step 0 to 10, then controlnet canny from step 10 to 20, then ends with no controlnet from step 20 to 30:
![image](https://github.com/comfyanonymous/ComfyUI/assets/34081873/71c564f7-3232-4e8c-9aa9-f67f776afc27)

For the Webui nodes, I'm using the A1111 Webui extension for ComfyUI. 

Note that it's possible to modify this workflow to change the starting/ending step of the 3 samples with 2 sliders, it's just that I don't have the proper number nodes for that:
![image](https://github.com/comfyanonymous/ComfyUI/assets/34081873/84b0a202-e493-4bf0-bfe9-e553bc60d34a)
![image](https://github.com/comfyanonymous/ComfyUI/assets/34081873/329f5881-5b8c-4836-8712-268ba4ca6fe6)

Here's the workflow if you want to play with it (it's zipped because github doesn't allow sharing json files):
[workflow (29).zip](https://github.com/comfyanonymous/ComfyUI/files/12160687/workflow.29.zip)

**Reply by notfood** [reply]
As per https://github.com/comfyanonymous/ComfyUI/commit/7ff14b62f87559938ab9e468c4e232201b8553ef you no longer need that complexity. Just add ControlNetApplyAdvanced and set start/end

**Reply by PladsElsker** [reply]
that's neat
